### ISSUE TITLE
Training Aborts with 1 checkpoint and WARNING:tensorflow:Gradients do not exist for variables ['top_bn/gamma:0', 'top_bn/beta:0'] when minimizing the loss. W0206 04:50:37.530761 140648355514112 utils.py:83] Gradients do not exist for variables ['top_bn/gamma:0', 'top_bn/beta:0'] when minimizing the loss.

### ISSUE THREAD

**Issue by nikhilparmar** [TannerGilbert/Tensorflow-Object-Detection-API-Train-Model#22]
I am running your Colab file on my Colab Pro account with 32GB and GPU enable.I have set the following parameters
```
batch_size = 1
num_steps = 8
num_eval_steps = 1
```

 Its running fine till the creation of 1st checkpoint, but after that it just stops with this on the output cell. 

```
W0206 04:50:15.624153 140651468793728 util.py:169] A checkpoint was restored (e.g. tf.train.Checkpoint.restore or tf.keras.Model.load_weights) but not all checkpointed values were used. See above for specific issues. Use expect_partial() on the load status object, e.g. tf.train.Checkpoint.restore(...).expect_partial(), to silence these warnings, or use assert_consumed() to make the check explicit. See https://www.tensorflow.org/guide/checkpoint#loading_mechanics for details.
WARNING:tensorflow:Gradients do not exist for variables ['top_bn/gamma:0', 'top_bn/beta:0'] when minimizing the loss.
W0206 04:50:25.243972 140648355514112 utils.py:83] Gradients do not exist for variables ['top_bn/gamma:0', 'top_bn/beta:0'] when minimizing the loss.
WARNING:tensorflow:From /usr/local/lib/python3.6/dist-packages/tensorflow/python/util/deprecation.py:605: calling map_fn_v2 (from tensorflow.python.ops.map_fn) with dtype is deprecated and will be removed in a future version.
Instructions for updating:
Use fn_output_signature instead
W0206 04:50:26.765114 140648355514112 deprecation.py:537] From /usr/local/lib/python3.6/dist-packages/tensorflow/python/util/deprecation.py:605: calling map_fn_v2 (from tensorflow.python.ops.map_fn) with dtype is deprecated and will be removed in a future version.
Instructions for updating:
Use fn_output_signature instead
WARNING:tensorflow:Gradients do not exist for variables ['top_bn/gamma:0', 'top_bn/beta:0'] when minimizing the loss.
W0206 04:50:37.530761 140648355514112 utils.py:83] Gradients do not exist for variables ['top_bn/gamma:0', 'top_bn/beta:0'] when minimizing the loss.

```
Please help

**Comment by TannerGilbert** [comment]
The `num_steps` parameter specifies how many images the model goes through when training (it's not epochs). Therefore you'll have to set the value a lot higher than 8. I set `num_steps` to 5000 and `num_eval_steps` to 1000 for training my microcontroller detector.

**Comment by nikhilparmar** [comment]
Okay will increase 

> The `num_steps` parameter specifies how many images the model goes through when training (it's not epochs). Therefore you'll have to set the value a lot higher than 8. I set `num_steps` to 5000 and `num_eval_steps` to 1000 for training my microcontroller detector.

Okay, so do you mean that its breaking in between because of low `num_steps` ? 
Now I set it to 80, now its taking more time, but hasn't gone ahead to check point 2. 
Also please let me know how much time it took for you to train with `num_steps=5000` and `num_eval_steps=1000`, it will give me a rough idea.

**Comment by TannerGilbert** [comment]
Sorry for the late reply. It seems like training isn't even starting. Else if would print something like:
```
INFO:tensorflow:Step 100 per-step time 0.963s loss=1.618
I0718 15:58:53.005567 139647596537728 model_lib_v2.py:635] Step 100 per-step time 0.963s loss=1.618
INFO:tensorflow:Step 200 per-step time 0.973s loss=0.983
I0718 16:00:31.596000 139647596537728 model_lib_v2.py:635] Step 200 per-step time 0.973s loss=0.983
INFO:tensorflow:Step 300 per-step time 0.948s loss=0.584
I0718 16:02:10.137982 139647596537728 model_lib_v2.py:635] Step 300 per-step time 0.948s loss=0.584
INFO:tensorflow:Step 400 per-step time 1.031s loss=0.462
...
```

Maybe try running it in Google Colab instead of running it globally. You can use my [Tensorflow_2_Object_Detection_Train_model.ipynb](https://github.com/TannerGilbert/Tensorflow-Object-Detection-API-Train-Model/blob/master/Tensorflow_2_Object_Detection_Train_model.ipynb) to get started.

**Comment by nikhilparmar** [comment]
I was running it on Colab Pro. Yes, training was not started. The issue is resolved now. The reason it was breaking on Colab is because the Tensorflow was not detecting and running on GPU even though GPU was enabled in Colab.
So, if anyone faces this issue, then one should check whether Tensorflow is running on GPU or not. Simplest solution is to Factory Reset Runtime, then Colab will reassign the GPU, then check, TF will now detect the GPU.
After all this, it ran smoothly.
Thanks

**Comment by nikhilparmar** [comment]
> The `num_steps` parameter specifies how many images the model goes through when training (it's not epochs). Therefore you'll have to set the value a lot higher than 8. I set `num_steps` to 5000 and `num_eval_steps` to 1000 for training my microcontroller detector.

Apologies for the noob questions, I have few doubts in this. 
1. After how many steps it saves a Checkpoint ?
2. How to set Epoch count, if we plan to run training for 5000 Epochs ?
3. what is num_eval_steps and why do we need it ?

Kindly help as it will clear my and others doubts.
Thanks

**Comment by TannerGilbert** [comment]
1. Checkpoints are saved every few minutes, but I don't know the exact time
2. The object detection API uses steps instead of epochs to define how long a model should be trained. First, they might seem completely different, but actually, it's quite simple to convert from one to another. An epoch simply tells you how often to go over all images, while num_steps tells you how many images you want to go over. So if you want to train for 50 epochs, you need to use 50*num_images steps.
3. The num_eval_steps is used to specify how often to evaluate the model

**Comment by nikhilparmar** [comment]
> 1. Checkpoints are saved every few minutes, but I don't know the exact time
> 2. The object detection API uses steps instead of epochs to define how long a model should be trained. First, they might seem completely different, but actually, it's quite simple to convert from one to another. An epoch simply tells you how often to go over all images, while num_steps tells you how many images you want to go over. So if you want to train for 50 epochs, you need to use 50*num_images steps.
> 3. The num_eval_steps is used to specify how often to evaluate the model

Thank you for the clear explanation.
1. If checkpoints are saved after few minutes, how to get the best checkpoint. I mean, support I train for 5000 num steps, and i observe that at around 3000 the accuracy/loss is best and then no improvement is observed or may be its getting worst, then how to pick this best performing checkpoint ? 
3. So is num_eval_steps =1000, then it means that after every 1000, the model is eval against validation/test set ?

Thanks.

**Comment by nikhilparmar** [comment]
> > 1. Checkpoints are saved every few minutes, but I don't know the exact time
> > 2. The object detection API uses steps instead of epochs to define how long a model should be trained. First, they might seem completely different, but actually, it's quite simple to convert from one to another. An epoch simply tells you how often to go over all images, while num_steps tells you how many images you want to go over. So if you want to train for 50 epochs, you need to use 50*num_images steps.
> > 3. The num_eval_steps is used to specify how often to evaluate the model
> 
> Thank you for the clear explanation.
> 
> 1. If checkpoints are saved after few minutes, how to get the best checkpoint. I mean, support I train for 5000 num steps, and i observe that at around 3000 the accuracy/loss is best and then no improvement is observed or may be its getting worst, then how to pick this best performing checkpoint ?
> 2. So is num_eval_steps =1000, then it means that after every 1000, the model is eval against validation/test set ?
> 
> Thanks.

@TannerGilbert Sir, it will really helpful for me if you can just answer my this query. As I am fully working on your code, and I am stuck at these questions. Only you can help me on this.

Thanks
